File: Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *merge(ListNode *p1, ListNode *p2) {
        ListNode *p, *ph ;
        ph = p = new ListNode(-1);
        while (p1 != NULL && p2 != NULL) {
            if (p1->val < p2->val) {
                p->next = p1;
                p = p1;
                p1 = p1->next;
            } else {
                p->next = p2;
                p = p2;
                p2 = p2->next;
            }
        }
        p->next = (p1 == NULL) ? p2 : p1;
        
        p = ph;
        ph = ph->next;
        free(p);
        
        return ph;
    }
    ListNode *mergeSort(ListNode *p, int len) {
        if (len == 1) {
            p->next = NULL;
            return p;
        }
        
        ListNode *pmid = p;
        for (int i = 0; i < len/2; i++) {
            pmid = pmid->next;
        }
        
        ListNode *p1 = mergeSort(p, len/2);
        ListNode *p2 = mergeSort(pmid, len - len/2);
        
        return merge(p1, p2);
    }
    ListNode *sortList(ListNode *head) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        if (head == 0) return NULL;
        
        int len = 0;
        ListNode *p = head;
        while (p != NULL) {
            p = p->next;
            ++len;
        }
        return mergeSort(head, len);
    }
};

File: Sort List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode SortList(ListNode head) {
        if (head == null) {
            return head;
        }
        
        var len = 0;
        var p = head;
        while (p != null) {
            p = p.next;
            len++;
        }
        
        
        return MergeSort(head, len);
    }
    
    private ListNode MergeSort(ListNode head, int len) {
        if (len == 1) {
            head.next = null;
            return head;
        }
        
        var mid = head;
        for (var i = 0; i < len / 2; i++) {
            mid = mid.next;
        }
        
        var p1 = MergeSort(head, len / 2);
        var p2 = MergeSort(mid, len - len / 2);
        
        return Merge(p1, p2);
    }
    
    private ListNode Merge(ListNode p1, ListNode p2) {
        var head = new ListNode(-1);
        var p = head;
        while (p1 != null && p2 != null) {
            if (p1.val < p2.val) {
                p.next = p1;
                p1 = p1.next;
            } else {
                p.next = p2;
                p2 = p2.next;
            }
            p = p.next;
        }
        p.next = (p1 == null ? p2 : p1);
        
        p = head;
        head = head.next;
        p.next = null;
        
        return head;
    }
}
